catalog/cataloger_merge.go

```diff
@@ -128,128 +165,99 @@ func hasCommitDifferences(tx db.Tx, leftID, rightID int64) (bool, error) {
 	return hasCommitDifferences, nil
 }
 
-func formatMergeMessage(leftBranch string, rightBranch string) string {
-	return fmt.Sprintf("Merge '%s' into '%s'", leftBranch, rightBranch)
-}
-
-func (c *cataloger) doMergeByRelation(ctx context.Context, tx db.Tx, relation RelationType, leftID, rightID int64, committer string, msg string, metadata Metadata) (CommitID, error) {
-	nextCommitID, err := getNextCommitID(tx)
-	if err != nil {
-		return 0, err
+func mergeBatch(tx db.Tx, mergeBatch mergeBatchType, previousMaxCommitID, nextCommitID CommitID, rightID int64) error {
+	// collect changes to apply  on the branch
+	paths := make([]string, 0, MergeBatchSize)
+	ctidArray := make([]string, 0, MergeBatchSize)
+	var tombstonePaths []string
+	for _, diffRec := range mergeBatch {
+		if diffRec.DiffType == DifferenceTypeRemoved || diffRec.DiffType == DifferenceTypeChanged {
+			paths = append(paths, diffRec.Entry.Path)
+		}
+		if (diffRec.DiffType == DifferenceTypeAdded || diffRec.DiffType == DifferenceTypeChanged) &&
+			diffRec.EntryCtid != nil {
+			ctidArray = append(ctidArray, *diffRec.EntryCtid)
+		}
+		if diffRec.DiffType == DifferenceTypeRemoved && diffRec.parentTombstoneNeeded {
+			tombstonePaths = append(tombstonePaths, diffRec.Entry.Path)
+		}
 	}
-
-	// do the merge based on the relation
-	previousMaxCommitID, err := getLastCommitIDByBranchID(tx, rightID)
-	if err != nil {
-		return 0, err
+	// apply changes
+	if len(paths) > 0 {
+		// set entries that exist in the right branch as deleted by entries that were removed or changed
+		setMaxCommit := sq.Update("catalog_entries").
```

Comment:
use `psql`

Reply:
Then I can not use debugSqelizer - non-starter for me